consensus/parlia/parlia.go

```diff
@@ -713,25 +713,40 @@ func (p *Parlia) snapshot(chain consensus.ChainHeaderReader, number uint64, hash
 			}
 		}
 
-		// If we're at the genesis, snapshot the initial state.
-		if number == 0 {
-			checkpoint := chain.GetHeaderByNumber(number)
-			if checkpoint != nil {
-				// get checkpoint data
-				hash := checkpoint.Hash()
-
+		// If we're at the genesis, snapshot the initial state. Alternatively if we have
+		// piled up more headers than allowed to be reorged (chain reinit from a freezer),
+		// consider the checkpoint trusted and snapshot it.
+		// even BEP-341 enabled, an offset `defaultEpochLength/2` can ensure getting the right validators.
+		if number == 0 || (number%p.config.Epoch == defaultEpochLength/2 && (len(headers) > int(params.FullImmutabilityThreshold)/10)) {
```

Comment:
why need (len(headers) > int(params.FullImmutabilityThreshold)/10)? We store snapshot every 1024 block. Is ok (len(headers) > checkpointInterval) ?

Reply:
this follow https://github.com/rjl493456442/go-ethereum/commit/536b3b416c6ff53ea11a0d29dcc351a6d7919901

Reply:
we store snapshot every 1024 block.  The stored snapshot can work should work In most cases.
the logic added here is used in an abnormal scene, so we know some thing wrong happed with storing snapshot.